src/util/AsioHelpers.h

```diff
@@ -1,65 +1,106 @@
+      // exception_ptr and the return value as the second argument.
+      if constexpr (isVoid) {
+        function_();
+        callHandler(nullptr);
+      } else {
+        callHandler(nullptr, function_());
+      }
+    } catch (...) {
+      // If `function_()` throws, we propagate the exception to the
+      // handler, and have no return value.
+      if constexpr (isVoid) {
+        callHandler(std::current_exception());
+      } else {
+        // Unfortunately we need to pass a `Value{}` even in the case of an
+        // error. This is due to a limitation in Boost::Asio.
+        callHandler(std::current_exception(), Value{});
+      }
+    }
   }
-  AD_CORRECTNESS_CHECK(exceptionPtr);
-  std::rethrow_exception(exceptionPtr);
-}
+};
+// Explicit deduction guides, we need objects and not references as the template
+// parameters.
+template <typename Executor, std::invocable Function, typename Handler>
+CallFunctionAndPassToHandler(Executor&&, Function&&, Handler&&)
+    -> CallFunctionAndPassToHandler<
+        std::decay_t<Executor>, std::decay_t<Function>, std::decay_t<Handler>>;
+}  // namespace detail
 
-// _____________________________________________________________________________
+// Run the `function` on the `executor` (e.g. a strand for synchronization or
+// the executor of a thread pool). As soon as the function has completed, the
+// completion handler which is obtained from the `completionToken` is posted to
+// its associated executor (which might be different from the `executor`).
+template <typename Executor, typename CompletionToken, std::invocable Function>
+requires std::is_default_constructible_v<std::invoke_result_t<Function>> ||
+         std::is_void_v<std::invoke_result_t<Function>>
+auto runFunctionOnExecutor(Executor executor, Function function,
+                           CompletionToken& completionToken) {
+  using Value = std::invoke_result_t<Function>;
+  static constexpr bool isVoid = std::is_void_v<Value>;
 
-/// Helper function that ensures that co_await resumes on the executor
-/// this coroutine was co_spawned on. Overload for void.
-inline net::awaitable<void> resumeOnOriginalExecutor(
-    net::awaitable<void> awaitable) {
-  std::exception_ptr exceptionPtr;
-  try {
-    co_await std::move(awaitable);
-  } catch (...) {
-    exceptionPtr = std::current_exception();
-  }
-  if ((co_await net::this_coro::cancellation_state).cancelled() ==
-      net::cancellation_type::none) {
-    // use_awaitable always resumes the coroutine on the executor the coroutine
-    // was co_spawned on
-    co_await net::post(net::use_awaitable);
-  }
-  if (exceptionPtr) {
-    std::rethrow_exception(exceptionPtr);
-  }
+  // Obtain the call signature for the handler.
+  [[maybe_unused]] auto getSignature = []() {
+    if constexpr (isVoid) {
+      return std::type_identity<void(std::exception_ptr)>{};
+    } else {
+      return std::type_identity<void(std::exception_ptr, Value)>{};
```

Comment:
Even better than using `std::type_identity` and then calling the default constructor you should use `std::declval<void(std::exception_ptr)>()` here which doesn't compile in an evaluated context. Then you can just decltype the return type of this helper lambda.
Then again you could also just write
```cpp
using Signature = std::conditional_t<isVoid, void(std::exception_ptr), void(std::exception_ptr, Value)>;
```
not sure if that's more readable though.